### PR TITLE
[FIX] Concordance: selection settings

### LINKED ISSUE #217: Concordanes Settings
##### Expected behavior
Concordanes widget uses settings. 

##### Actual behavior
It does not yet.

### PR DESCRIPTION
**Pull request by jerneju**
##### Issue
Fixes #217.

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: orangecontrib/text/widgets/owconcordance.py [hunk 1]
```diff
@@ -151,10 +151,14 @@ class OWConcordance(OWWidget):
...
 
+    settingsHandler = PerfectDomainContextHandler(
+        match_values=PerfectDomainContextHandler.MATCH_VALUES_ALL
+    )
+
     autocommit = Setting(True)
     context_width = Setting(5)
-    word = Setting("")
-    # TODO Set selection settings.
+    word = ContextSetting("b", exclude_metas=False)
```
Comment (ajdapretnar) [inline orangecontrib/text/widgets/owconcordance.py]: Why is "b" the default? Please set it to an empty string.

File: orangecontrib/text/widgets/owconcordance.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ def flags(self, _):
         return Qt.ItemIsEnabled | Qt.ItemIsSelectable
 
     def rowCount(self, parent=None, *args, **kwargs):
-        return 0 if parent.isValid() or self.word_index is None \
+        return 0 if parent and parent.isValid() or self.word_index is None \
```
Comment (ajdapretnar) [inline orangecontrib/text/widgets/owconcordance.py]: This still fails if I select some documents in Concordance and then remove the data. It is fixed in #232.

File: orangecontrib/text/widgets/owconcordance.py [hunk 3]
```diff
@@ -223,11 +227,15 @@ def set_width(self):
                 QItemSelectionModel.SelectCurrent | QItemSelectionModel.Rows)
 
     def set_corpus(self, data=None):
+        self.closeContext()
+
```
Comment (ajdapretnar) [inline orangecontrib/text/widgets/owconcordance.py]: While at it, would it not make sense to remove the empty line? Or is this intentional?

File: orangecontrib/text/widgets/owconcordance.py [hunk 4]
```diff
@@ -223,11 +227,15 @@ def set_width(self):
...
         self.corpus = data
         if data is not None and not isinstance(data, Corpus):
             self.corpus = Corpus.from_table(data.domain, data)
         self.model.set_corpus(self.corpus)
-        self.update_widget()
+        self.openContext(self.corpus)
+
```
Comment (ajdapretnar) [inline orangecontrib/text/widgets/owconcordance.py]: Same as above.

File: orangecontrib/text/widgets/owconcordance.py (hunk 1 above)
Comment (jerneju) [inline orangecontrib/text/widgets/owconcordance.py]: Well the widget is still WIP, I have not done the cleaning yet. Anyway thanks for the comments.

**Comment by ajdapretnar** [comment]
Doh, sorry, I didn't see the tag. 🙈 Silly me.
I was just in a hurry to merge EVERYTHING!! 🤸‍♂️ 🤣

**Comment by ajdapretnar** [comment]
So is this done or do we wait for similar fixes as in #241 (list issue)?
+ I think this needs a rebase?

**Comment by jerneju** [comment]
Rebased.

**Comment by ajdapretnar** [comment]
@nikicc Please check as well and merge. :)

File: orangecontrib/text/widgets/owconcordance.py [hunk 5]
```diff
@@ -235,12 +239,32 @@ def set_width(self):
             self.conc_view.selectionModel().select(sel,
                 QItemSelectionModel.SelectCurrent | QItemSelectionModel.Rows)
 
+    def get_selection(self):
+        sel = self.conc_view.selectionModel().selection()
+        selection = []
```
Comment (nikicc) [inline orangecontrib/text/widgets/owconcordance.py]: Why not just: 
```py
selection = sorted(set(cell.row() for cell in sel.indexes()))
```

File: orangecontrib/text/widgets/owconcordance.py [hunk 6]
```diff
@@ -256,7 +280,7 @@ def set_word_from_input(self, topic):
     def set_word(self):
         self.model.set_word(self.word)
         self.update_widget()
-        self.commit()
+        self.set_selection()
```
Comment (nikicc) [inline orangecontrib/text/widgets/owconcordance.py]: Why this? Currently, if I am inputing different words on input the selection persists which is kind on unintuitive?

**Comment by jerneju** [comment]
Rows selection not finished yet. Waiting for`DataHashContextHandler`to be implemented. Similar problem as in #255 .

**Comment by nikicc** [comment]
We should probably do the same for #256 to avoid problems with selection being stored in context settings.

**Comment by ajdapretnar** [comment]
@jerneju What is the state of this PR? Is it still WIP? Can we expect to merge soon?

**Comment by jerneju** [comment]
@ajdapretnar: It is still impossible to implement this since we cannot do it without proper context handler. See my comment above.

**Comment by ajdapretnar** [comment]
Ok, next step. Is anyone working on the context handler? If not, should we assign? It would be great to finish this before the workshop.

File: orangecontrib/text/widgets/owconcordance.py [hunk 7]
```diff
@@ -268,6 +283,13 @@ def set_word(self):
         self.update_widget()
         self.commit()
 
+    def handleNewSignals(self):
+        selection = self.selection
+        if self.corpus:
+            self._set_selection(selection)
+        else:
+            self._set_selection([0])
```
Comment (nikicc) [inline orangecontrib/text/widgets/owconcordance.py]: Same here, why do we set `[0]` as selection if there is no corpus?

File: orangecontrib/text/widgets/owconcordance.py [hunk 8]
```diff
@@ -268,6 +283,13 @@ def set_word(self):
         self.update_widget()
         self.commit()
 
+    def handleNewSignals(self):
+        selection = self.selection
+        if self.corpus:
+            self._set_selection(selection)
```
Comment (nikicc) [inline orangecontrib/text/widgets/owconcordance.py]: `self._set_selection(self.selection)`?

File: orangecontrib/text/widgets/owconcordance.py [hunk 9]
```diff
@@ -240,6 +240,21 @@ def set_width(self):
             self.conc_view.selectionModel().select(sel,
                 QItemSelectionModel.SelectCurrent | QItemSelectionModel.Rows)
 
+    def _save_selection(self):
+        sel = self.conc_view.selectionModel().selection()
+        selection = sorted(set(cell.row() for cell in sel.indexes()))
+        if selection:
```
Comment (nikicc) [inline orangecontrib/text/widgets/owconcordance.py]: Why this if?

File: orangecontrib/text/widgets/owconcordance.py [hunk 10]
```diff
@@ -173,7 +173,7 @@ class Outputs:
     autocommit = Setting(True)
     context_width = Setting(5)
     word = ContextSetting("", exclude_metas=False)
-    # TODO Set selection settings (DataHashContextHandler)
+    selection = Setting([0], schema_only=True)
```
[the remark this answers is not in the document]
Comment (ajdapretnar) [reply orangecontrib/text/widgets/owconcordance.py]: Hm... Perhaps you are right. Selecting the first item only makes sense in Corpus Viewer, but not in Concordance. I'm sorry, this is probably my fault as I said to make it the same as CV.

[the remark this answers is not in the document]
Comment (jerneju) [reply orangecontrib/text/widgets/owconcordance.py]: Ok, let it be empty.

File: orangecontrib/text/widgets/owconcordance.py (hunk 8 above)
Comment (jerneju) [reply orangecontrib/text/widgets/owconcordance.py]: Fixed.

File: orangecontrib/text/widgets/owconcordance.py [hunk 11]
```diff
@@ -259,17 +255,21 @@ def _set_selection(self, selection):
...
             self.openContext(self.corpus)
+
+        self.model.set_corpus(self.corpus)
         self.set_word()
 
     @Inputs.query_word
     def set_word_from_input(self, topic):
         self.Warning.multiple_words_on_input.clear()
+        if self.is_word_on_input:   # word changed, clear selection
+            self.selected_rows = []
```
Comment (jerneju) [inline orangecontrib/text/widgets/owconcordance.py]: Why `if`?

Comment (nikicc) [reply orangecontrib/text/widgets/owconcordance.py]: So it only resets when we change the input word from one topic to another and not also when we reload the scema and topic is passed on input.

I think that without this when reloading selection would be reset before `handleNewSignals` would have the chance to set it.

File: orangecontrib/text/widgets/owconcordance.py [hunk 12]
```diff
@@ -259,17 +255,21 @@ def _set_selection(self, selection):
     def set_corpus(self, data=None):
         self.closeContext()
         self.corpus = data
-        if data is not None and not isinstance(data, Corpus):
-            self.corpus = Corpus.from_table(data.domain, data)
```
Comment (jerneju) [inline orangecontrib/text/widgets/owconcordance.py]: I think these two lines should not have been removed. Now the test fails.

**Comment by nikicc** [comment]
@jerneju I removed the test that doesn't apply anymore. Should be fine now.

**Pull request merged**